File: Behavioral/Interpreter/Python/interpreter_client.py
# interpreter_client.py

from number_expression import NumberExpression
from add_expression import AddExpression
from subtract_expression import SubtractExpression

def parse_expression(expression):
  stack = []
  tokens = expression.split()

  for token in tokens:
    if token == "+":
      right = stack.pop()
      left = stack.pop()
      stack.append(AddExpression(left, right))
    elif token == "-":
      right = stack.pop()
      left = stack.pop()
      stack.append(SubtractExpression(left, right))
    else:
      stack.append(NumberExpression(int(token)))

  return stack.pop()

File: Behavioral/Interpreter/Java/InterpreterClient.java
// InterpreterClient.java

import java.util.Stack;

public class InterpreterClient {
  public static Expression parseExpression(String expression) {
    Stack<Expression> stack = new Stack<>();

    String[] tokens = expression.split(" ");
    for (String token : tokens) {
      if (token.equals("+")) {
        Expression right = stack.pop();
        Expression left = stack.pop();
        stack.push(new AddExpression(left, right));
      } else if (token.equals("-")) {
        Expression right = stack.pop();
        Expression left = stack.pop();
        stack.push(new SubtractExpression(left, right));
      } else {
        stack.push(new NumberExpression(Integer.parseInt(token)));
      }
    }
    return stack.pop();
  }
}
